File: 3rd-semester/object-oriented-programming/2nd-assignment/csharp/src/Services/Registry.cs
using Microsoft.Extensions.Logging;
using SurveillanceGui.Models;

namespace SurveillanceGui.Services;

public class Registry
{
    private static readonly IReadOnlySet<string> SuspiciousKeywords =
        new HashSet<string> { "Bomb", "Attack", "Gun", "Explosives" };

    private readonly ILogger<Registry>   _logger;
    private readonly List<Suspect>       _suspects       = [];
    private readonly List<Communication> _communications = [];

    public Registry(ILogger<Registry> logger) => _logger = logger;

    public IReadOnlyList<Suspect> Suspects => _suspects.AsReadOnly();

    public void AddSuspect(Suspect suspect)
    { _suspects.Add(suspect); _logger.LogInformation("Registered: {Name}", suspect.Name); }

    public void AddCommunication(Communication communication)
    {
        _communications.Add(communication);
        var s1 = FindByNumber(communication.SenderNumber);
        var s2 = FindByNumber(communication.ReceiverNumber);
        if (s1 is not null && s2 is not null && s1 != s2)
        { s1.AddCollaborator(s2); s2.AddCollaborator(s1); }
    }

    public Suspect? GetSuspectByName(string name) =>
        _suspects.FirstOrDefault(s => s.Name.Equals(name, StringComparison.OrdinalIgnoreCase));

    public Suspect? GetSuspectWithMostPartners() => _suspects.MaxBy(s => s.CollaboratorCount);

    public IReadOnlyList<Sms> GetSuspiciousMessagesBetween(string number1, string number2) =>
        _communications
            .Where(c => Involves(c, number1, number2) && !string.IsNullOrEmpty(c.GetSmsContent()))
            .Where(c => SuspiciousKeywords.Any(kw => c.GetSmsContent().Contains(kw)))
            .Cast<Sms>().ToList();

    public IReadOnlyList<Suspect> GetSuspectsFromCountry(string country) =>
        _suspects.Where(s => s.Country == country).ToList();

    private Suspect? FindByNumber(string number) =>
        _suspects.FirstOrDefault(s => s.PhoneNumbers.Contains(number));

    private static bool Involves(Communication c, string n1, string n2) =>
        (c.SenderNumber == n1 && c.ReceiverNumber == n2) ||
        (c.SenderNumber == n2 && c.ReceiverNumber == n1);
}

File: 3rd-semester/object-oriented-programming/2nd-assignment/java/src/main/java/com/university/oop/surveillance/service/Registry.java
package com.university.oop.surveillance.service;

import com.university.oop.surveillance.model.Communication;
import com.university.oop.surveillance.model.PhoneCall;
import com.university.oop.surveillance.model.Sms;
import com.university.oop.surveillance.model.Suspect;

import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Collectors;

/** Central registry managing suspects and their communications. */
public class Registry {

    private static final Logger LOGGER = Logger.getLogger(Registry.class.getName());
    private static final Set<String> SUSPICIOUS_KEYWORDS =
            Set.of("Bomb", "Attack", "Gun", "Explosives");

    private final List<Suspect>       suspects       = new ArrayList<>();
    private final List<Communication> communications = new ArrayList<>();

    public void addSuspect(Suspect suspect) { suspects.add(suspect); }

    public List<Suspect> getSuspects() { return Collections.unmodifiableList(suspects); }

    public void addCommunication(Communication communication) {
        communications.add(communication);
        Optional<Suspect> s1 = findByNumber(communication.getSenderNumber());
        Optional<Suspect> s2 = findByNumber(communication.getReceiverNumber());
        if (s1.isPresent() && s2.isPresent() && !s1.equals(s2)) {
            s1.get().addCollaborator(s2.get());
            s2.get().addCollaborator(s1.get());
        }
    }

    public Optional<Suspect> getSuspectWithMostPartners() {
        return suspects.stream().max(Comparator.comparingInt(Suspect::getCollaboratorCount));
    }

    public Optional<PhoneCall> getLongestPhoneCallBetween(String number1, String number2) {
        return communications.stream()
                .filter(c -> c.getDuration() > 0 && involves(c, number1, number2))
                .map(c -> (PhoneCall) c)
                .max(Comparator.comparingInt(PhoneCall::getDuration));
    }

    public List<Sms> getSuspiciousMessagesBetween(String number1, String number2) {
        return communications.stream()
                .filter(c -> involves(c, number1, number2))
                .filter(c -> !c.getSmsContent().isEmpty())
                .filter(c -> SUSPICIOUS_KEYWORDS.stream().anyMatch(c.getSmsContent()::contains))
                .map(c -> (Sms) c)
                .collect(Collectors.toList());
    }

    private Optional<Suspect> findByNumber(String number) {
        return suspects.stream().filter(s -> s.getPhoneNumbers().contains(number)).findFirst();
    }

    private boolean involves(Communication c, String n1, String n2) {
        String s = c.getSenderNumber(), r = c.getReceiverNumber();
        return (s.equals(n1) && r.equals(n2)) || (s.equals(n2) && r.equals(n1));
    }
}
